### PR TITLE
upgrade mypy and fix issues [sc-15688]

### PR DESCRIPTION
**Pull request by alyiwang**
### 🤔 Why?

To keep our code typing more robust

### 🤓 What?

- upgrade mypy to 0.991
- fix 100+ issues and bugs
- format some code

### 🧪 Tested?

all unit tests pass
run snowflake and bigquery crawlers locally and verified the results

### CODE REVIEW

File: metaphor/bigquery/config.py [hunk 1]
```diff
@@ -56,15 +55,15 @@ class BigQueryQueryLogConfig:
...
+    project_id: str
+
     # Path to service account's JSON key file
     key_path: Optional[str] = None
 
     # The credentials from the BigQuery JSON key file
     credentials: Optional[BigQueryCredentials] = None
 
-    # Project ID to use. Use the service account's default project if not set
-    project_id: Optional[str] = None
```
Comment (mars-lan) [inline metaphor/bigquery/config.py:66]: Why changing this to required?

Comment (alyiwang) [reply metaphor/bigquery/config.py:66]: This is mentioned as a required config in https://github.com/MetaphorData/connectors/tree/main/metaphor/bigquery#required-configurations

Comment (mars-lan) [reply metaphor/bigquery/config.py:66]: SG

File: tests/power_bi/test_parse_power_query.py [hunk 2]
```diff
@@ -5,14 +5,14 @@
 
 def test_parse_dataset_redshift():
     exp = 'let\n    Source = AmazonRedshift.Database("url:5439", "db" ),\n    public = Source{[Name="public"]}[Data],\n    table1 = public{[Name="table"]}[Data]\nin\n    table1'
-    assert PowerQueryParser.parse_source_datasets(exp) == [
+    assert PowerQueryParser.parse_source_datasets(exp, None) == [
```
Comment (mars-lan) [inline tests/power_bi/test_parse_power_query.py:8]: Do we need to explicitly specify optional arguments as `None`?

File: tests/tableau/test_extractor.py [hunk 3]
```diff
@@ -141,7 +141,7 @@ def test_parse_dataset_id():
             name="name",
             fullName="db.schema.table",
             database=Database(name="db", connectionType="redshift"),
-            schema_=None,
+            schema="foo",
```
Comment (mars-lan) [inline tests/tableau/test_extractor.py:144]: Why changing these from `None` to `foo`?

File: tests/power_bi/test_parse_power_query.py (hunk 2 above)
Comment (alyiwang) [reply tests/power_bi/test_parse_power_query.py:8]: We can add default value `None` to the input param

File: metaphor/thought_spot/extractor.py [hunk 4]
```diff
@@ -453,6 +458,7 @@ def _populate_source_virtual_views(
                 )
                 for viz, *_ in charts
                 for column in viz.vizContent.columns
+                if column.referencedTableHeaders
```
Comment (mars-lan) [inline metaphor/thought_spot/extractor.py:461]: The complex comprehension is hard to enough to read in the first place, adding this `if` making it even worse. Also shouldn't it comes after the last for?

File: metaphor/tableau/extractor.py [hunk 5]
```diff
@@ -360,7 +364,12 @@ def _parse_upstream_datasets(
         )
 
     def _parse_dataset_id(self, table: DatabaseTable) -> Optional[EntityId]:
-        if None in (table.name, table.schema, table.fullName, table.database):
+        if (
```
Comment (mars-lan) [inline metaphor/tableau/extractor.py:367]: What was wrong with the previous code, which was more readable?

File: metaphor/snowflake/extractor.py [hunk 6]
```diff
@@ -130,7 +129,7 @@ async def extract(self) -> Collection[ENTITY_TYPES]:
             if self._query_log_lookback_days > 0:
                 self._fetch_query_logs()
 
-        datasets = self._datasets.values()
+        datasets = list(self._datasets.values())
```
Comment (mars-lan) [inline metaphor/snowflake/extractor.py:132]: I thought `dict.values()` is already a `list`?

File: tests/power_bi/test_parse_power_query.py (hunk 2 above)
Comment (mars-lan) [reply tests/power_bi/test_parse_power_query.py:8]: Yeah might be more pythonic that way.

File: tests/tableau/test_extractor.py (hunk 3 above)
Comment (alyiwang) [reply tests/tableau/test_extractor.py:144]: This is to test the case that with `fullname`, the `schema` won't be used.

File: metaphor/tableau/extractor.py (hunk 5 above)
Comment (alyiwang) [reply metaphor/tableau/extractor.py:367]: for `if None in (table.name, table.schema, table.fullName, table.database)`, mypy won't be able to tell that all of the terms are non-null. Has to write it the other way. 😞

File: metaphor/snowflake/extractor.py (hunk 6 above)
Comment (alyiwang) [reply metaphor/snowflake/extractor.py:132]: it's a view

File: metaphor/thought_spot/extractor.py (hunk 4 above)
Comment (alyiwang) [reply metaphor/thought_spot/extractor.py:461]: I personally find it pretty straightforward to read, the structure is the same as writing nested for loops. The `if` has to come after `column` but before `reference`, otherwise, mypy throws `None value is not iterable` for `column.referencedTableHeaders`

Comment (alyiwang) [reply metaphor/thought_spot/extractor.py:461]: Just have to visually imagine there are indents.

Comment (mars-lan) [reply metaphor/thought_spot/extractor.py:461]: Got it.

File: metaphor/tableau/extractor.py (hunk 5 above)
Comment (mars-lan) [reply metaphor/tableau/extractor.py:367]: Okay. In that case, using `is None` is probably safer than testing for falsey?

File: metaphor/snowflake/extractor.py [hunk 7]
```diff
@@ -244,7 +243,7 @@ def _fetch_columns(self, cursor: SnowflakeCursor, database: str) -> None:
     def _fetch_table_info(
         self, tables: Dict[str, DatasetInfo], is_shared_database: bool
     ) -> None:
-        dict_cursor = self._conn.cursor(DictCursor)
+        dict_cursor: DictCursor = self._conn.cursor(DictCursor)  # type: ignore
```
Comment (mars-lan) [inline metaphor/snowflake/extractor.py:246]: Why ignore?

File: metaphor/power_bi/power_query_parser.py [hunk 8]
```diff
@@ -217,7 +218,7 @@ def replacer(match: re.Match) -> str:
         power_query = re.sub(r"#\((cr|lf|tab)(,(cr|lf|tab))*\)", replacer, power_query)
 
         if "Value.NativeQuery(" in power_query:
-            return PowerQueryParser.parse_native_query(power_query, snowflake_account)
+            return PowerQueryParser._parse_native_query(power_query, snowflake_account)
 
         try:
             return [PowerQueryParser.parse_power_query(power_query)]
```
Comment (mars-lan) [inline metaphor/power_bi/power_query_parser.py:224]: Also mark `parse_power_query` as private for consistency?

File: metaphor/common/event_util.py [hunk 9]
```diff
@@ -5,7 +5,7 @@
 
 import fastjsonschema
 
-from metaphor import models
+from metaphor import models  # type: ignore
```
Comment (mars-lan) [inline metaphor/common/event_util.py:8]: Why ignore?

File: metaphor/snowflake/extractor.py (hunk 7 above)
Comment (alyiwang) [reply metaphor/snowflake/extractor.py:246]: `connection.cursor()` method has dynamic typing based on the given type param. But the method's returned type is always labeled `SnowflakeCursor`, and mypy throws `SnowflakeCursor` not assignable to `DictCursor`.

File: metaphor/common/event_util.py (hunk 9 above)
Comment (alyiwang) [reply metaphor/common/event_util.py:8]: Mypy: `model` is not a module under metaphor.
We use this import to load the schema json file for validation.

File: metaphor/power_bi/power_query_parser.py (hunk 8 above)
Comment (alyiwang) [reply metaphor/power_bi/power_query_parser.py:224]: updated

File: metaphor/tableau/extractor.py (hunk 5 above)
Comment (alyiwang) [reply metaphor/tableau/extractor.py:367]: falsy also test for an empty string which we don't want either.

**Comment by elic-eon** [comment]
https://mypy-lang.blogspot.com/2023/02/mypy-10-released.html Mypy 1.0 was released.

**Comment by alyiwang** [comment]
> https://mypy-lang.blogspot.com/2023/02/mypy-10-released.html Mypy 1.0 was released.

I tried that 😄 . But mypy 1.0 requires python 3.8+, and we are at 3.7+ now.

**Review by mars-lan — APPROVED** [review]
Thanks for fixing these!
